### PR TITLE
appcd-default-plugins v4

### PR DESCRIPTION
**Pull request by cb1kenobi**
# DO NOT MERGE until https://github.com/appcelerator/appc-daemon-plugins/pull/9 has been merged

- [x] Attempt to install the default plugins during `postinstall`
- [x] Need to test/fix when plugins dir already has plugins
- [x] Only go through install process if appcd-default-plugin's `package.json` has version that is not installed
- [x] Try to install default plugins at appcd install time by checking directory permissions
- [x] Only download plugins compatible with current platform

BREAKING CHANGE(appcd-default-plugins): Switched to installing default plugins at runtime to avoid user permissions.
chore(appcd-default-plugins): v4.0.0

https://jira.appcelerator.org/browse/DAEMON-280

`appcd status` shows the registered plugins:

```
Plugin       Version  Type      Path                                                                               Node.js  Status                                   Active/Total Requests
android      1.4.0    external  /Users/chris/.appcelerator/appcd/plugins/packages/@appcd/plugin-android/1.4.0      10.16.0  Inactive                                 0 / 0
android      2.0.0    external  /Users/chris/appc/appcd/appc-daemon-plugins/packages/plugin-android                10.16.0  Inactive                                 0 / 0
genymotion   1.4.0    external  /Users/chris/appc/appcd/appc-daemon-plugins/packages/plugin-genymotion             10.16.0  Inactive                                 0 / 0
ios          1.4.0    external  /Users/chris/.appcelerator/appcd/plugins/packages/@appcd/plugin-ios/1.4.0          10.16.0  Inactive                                 0 / 0
ios          2.0.0    external  /Users/chris/appc/appcd/appc-daemon-plugins/packages/plugin-ios                    10.16.0  Inactive                                 0 / 0
jdk          1.4.0    external  /Users/chris/appc/appcd/appc-daemon-plugins/packages/plugin-jdk                    10.16.0  Inactive                                 0 / 0
system-info  1.4.0    external  /Users/chris/.appcelerator/appcd/plugins/packages/@appcd/plugin-system-info/1.4.0  10.16.0  Inactive                                 0 / 0
system-info  2.0.0    external  /Users/chris/appc/appcd/appc-daemon-plugins/packages/plugin-system-info            10.16.0  Inactive                                 0 / 0
titanium     1.5.0    external  /Users/chris/appc/appcd/appc-daemon-plugins/packages/plugin-titanium               10.16.0  Inactive                                 0 / 0
windows      1.4.0    external  /Users/chris/appc/appcd/appc-daemon-plugins/packages/plugin-windows                10.16.0  Inactive: Unsupported platform "darwin"  0 / 0
```

### CODE REVIEW

File: packages/appcd-default-plugins/src/index.js [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+	} else {
+		logger.error('Unable to find lerna bin, skipping install default plugins');
+		return;
+	}
+
+	const packagesDir = path.join(pluginsDir, 'packages');
+	const invalidDir = path.join(pluginsDir, 'invalid');
+	const linksDir = path.join(os.homedir(), '.config', 'yarn', 'link');
+	const workspaces = [];
+	const { plugins } = fs.readJsonSync(path.resolve(__dirname, '..', 'package.json'));
```
Comment (ewanharris) [inline packages/appcd-default-plugins/src/index.js]: Maybe use the promise version other *Sync for this and other fs methods throughtout

Comment (cb1kenobi) [reply packages/appcd-default-plugins/src/index.js]: Fixed!

**Comment by cb1kenobi** [comment]
@ewanharris I refactored things and I think I've come up with a viable hack to get the default plugins installed. What do you think?

Also, any idea why Jenkins is not liking `appcd-logger`?

**Comment by ewanharris** [comment]
@cb1kenobi, I don't think it's Jenkins not liking it, it seems like a genuine bug in the dev flow with this change. To repro

1. Clean your repo `gulp clean --commit`
2. Run `yarn` or `lerna bootstrap`

And you'll get the error above, basically it tries to run the postinstall step after it's installed/linked all modules but **before** it's built the sources so 💥 . I guess maybe we need a short circuit?

**Comment by cb1kenobi** [comment]
Closing due to restructure.

**Pull request closed without merging**